### PR TITLE
Export database of Users

### PR DESCRIPTION
**Pull request by sreejakshetty**
#12

screenshot from 2016-06-06 16 41 34
screenshot from 2016-06-06 16 41 21

### CODE REVIEW

**Comment by sreejakshetty** [comment]
#12 tested on online server

**Comment by hpdang** [comment]
it doesn't work for me. Can you share the exported file you get? 

screenshot from 2016-06-07 11 43 49
screenshot from 2016-06-07 11 44 06

**Comment by sreejakshetty** [comment]
screenshot from 2016-06-07 15 41 39
screenshot from 2016-06-07 15 41 24
screenshot from 2016-06-07 15 40 53

**Comment by sreejakshetty** [comment]
I am getting the html code with the user data . I will check once why I am getting that error.They are getting inserted into the User table.

**Comment by hpdang** [comment]
@DishantK1807 I want to have this feature completed for Milestone 5, please test and finish it up.

**Comment by DishantK1807** [comment]
Will start working on it.

**Comment by DishantK1807** [comment]
@kamishettysreeja25 : You are using functions in your code that are deprecated: 
- mysql_connect()
- mysql_errno()
- mysql_error()
- mysql_fetch_row()
- mysql_field_name()
- mysql_num_fields()
- mysql_query()
- mysql_select_db()

the above-mentioned functions are deprecated and we cannot use them in our code. That's the reason we are getting HTML code along with the file.

**Comment by DishantK1807** [comment]
@kamishettysreeja25 : how did you get the screenshot of the database above?

**Comment by sreejakshetty** [comment]
@DishantK1807 Which screenshot are you talking about ?
I am able to extract user data but I am getting html code along with it . screenshot shows the user data.

**Comment by DishantK1807** [comment]
@kamishettysreeja25 : Did you changed the functions?

**Comment by DishantK1807** [comment]
@hpdang : I am not able to find possible error in the code. @hongquan :  can you suggest  me a way out?

**Comment by hongquan** [comment]
@kamishettysreeja25 Fix the issue of using deprecated `mysql_xx` functions first. Reused the things in _includes/mysqli_provider.php_. I cannot test because those `mysql_xx` don't work on my machine.

The _includes/mysqli_provider.php_ script already takes care of connecting to database, and return the `global $sql_connection` variable. You don't need to do this step again in your code.

**Comment by sreejakshetty** [comment]
@hongquan sure. I will work on this issue . I will try to use the functions in mysqli_provider.php and implement it .@DishantK1807  is also working on the same issue . Should we both work on same issue ?

**Comment by DishantK1807** [comment]
@hongquan : Changed those functions. But after changing them the exported file shows "Unable to load angeltypes" 
capture

**Comment by hongquan** [comment]
I recommend you guys to use PHP's built-in server to discover early the PHP error from the console.

php -s localhost 5000_101

@DishantK1807 : Why do you hard-code the MySQL user and password in `export_xls()`:

```
$Connect = sql_connect("localhost", "root", "","engelsystem")
```

What's about if the website is not setup with those value?

Please read my comment above, where I mention about reusing something in  _includes/mysqli_provider.php_.

Also, don't name variable with capitalize letter. Name it `$mything` instead of `$Mything` or `$myThing`.

**Comment by hongquan** [comment]
@kamishettysreeja25  Sister H.Phuc wants you guys to finish this soon, you guys can work together on it. Use private chat to decide who to write code, commit code.

**Comment by hongquan** [comment]
Try to fix the error which PHP reported:

```
[Fri Jun 17 13:35:53 2016] PHP Warning:  mysqli::query() expects parameter 1 to be string, array given in /home/quan/Works/FOSSASIA/engelsystem/includes/mysqli_provider.php on line 147
[Fri Jun 17 13:35:53 2016] PHP Notice:  Array to string conversion in /home/quan/Works/FOSSASIA/engelsystem/includes/mysqli_provider.php on line 151
[Fri Jun 17 13:35:53 2016] mysql_provider error: MySQL-query error: Array (0)
#1  sql_error(MySQL-query error: Array (0)
) called at [/home/quan/Works/FOSSASIA/engelsystem/includes/mysqli_provider.php:151]
```

**Comment by hongquan** [comment]
@DishantK1807 What is the purpose of the last 2 commits, b1be305, 302cf0e? They don't solve any thing and you still push to repo? Please, at least test on your localhost to make it somehow work, before pushing here. Don't pollute the Git history.

Before pushing, please make your commit neat. Those two actually can combine to one. Learn and  use `git rebase -i` to do. Note:  Do `rebase` only on the commits which you haven't pushed to repo.

Those 2 commit even introduce 1 more error:

```
[Fri Jun 17 14:56:18 2016] PHP Notice:  Undefined variable: sql_connection in /home/quan/Works/FOSSASIA/engelsystem/includes/pages/admin_import.php on line 395
```

Please try to test on your localhost first. Use PHP's builtin server to see error early!

**Comment by DishantK1807** [comment]
Sreeja and I tested the new code on our localservers for errors. The exported file still contains HTML code along with the data.

**Comment by hongquan** [comment]
That HTML is from this code in _public/index.php_:

``` php
echo template_render('../templates/layout.html', array(
    'theme' => isset($user) ? $user['color'] : $default_theme,
    'title' => $title,
    'atom_link' => ($p == 'news' || $p == 'user_meetings') ? '<link href="' . page_link_to('atom') . (($p == 'user_meetings') ? '&amp;meetings=1' : '') . '&amp;key=' . $user['api_key'] . '" type="application/atom+xml" rel="alternate" title="Atom Feed">' : '',
    'menu' => make_menu(),
    'content' => msg() . $content,
    'header_toolbar' => header_toolbar(),
    'faq_url' => $faq_url,
    'contact_email' => $contact_email,
    'locale' => locale() 
));
```

this code is always run and it inserts HTML content to every HTTP response. You have to add some logic to exclude it from your download page.

And @DishantK1807 , you violate coding style again, why all lines in `export_xls()` out of indentation? And why the file name _Webinfopen.csv_? Is it related any to "export user" feature? Did you just copy and paste from somewhere and too lazy to review what it is doing actually?

**Comment by DishantK1807** [comment]
@hongquan : this code was just to know what the problem was and I forgot about the indentation, sorry for that, and now that I've figured out what the actual problem is, I'll work on finding a solution to this problem and make the code right, according to the system.

**Comment by DishantK1807** [comment]
capture

User database is now exported without the HTML code.

Tested on the online server.

**Comment by DishantK1807** [comment]
@hongquan: Should I make a new PR? as this PR contains many commits that are unnecessary, or we can merge this PR only?

**Comment by hongquan** [comment]
Your last commit should be named "Fix: HTML is inserted in export result". I saw many commits with the same title "export user database".

No need to make new PR. You just need to pay attention to the commits next time.

Please make the file name more specific than "export_data.csv". Which data?

Out of indentation:

selection_104

**Comment by DishantK1807** [comment]
@hongquan : Updated.

**Comment by hongquan** [comment]
Our coding convention is: "Indentation width is 2 spaces, not tab".

File: includes/pages/admin_import.php [hunk 1]
```diff
@@ -81,69 +81,83 @@ function admin_import() {
           error(_('Please provide some data.'));
         }
       }
-      
+      if(isset($_REQUEST['download'])){
```
Comment (hongquan) [inline includes/pages/admin_import.php:84]: whitespace

**Comment by hongquan** [comment]
@DishantK1807 : Many issues:
- Why you report to browser that the file to be delivered is a XLS file, while the real content is CSV?
- Wrong header data
  selection_106
- You code causes creating a file _export_users_data.csv_ in _public_ folder. What if many users are using the _export_ feature nearly at the same time? They all write to the same file. How the file will be if many people write to it at the same time?

selection_107

**Comment by DishantK1807** [comment]
@hongquan : 

> You code causes creating a file export_users_data.csv in public folder. What if many users are using the export feature nearly at the same time? They all write to the same file. How the file will be if many people write to it at the same time?

I am creating a new file on the server so that the HTML content is not inserted into the exported file, that file is saved in the public directory. Only admin can download the user data, no other user is allowed to do so, many people cannot write to it at the same time. Only admin can. And each time the data is exported, the previous file will be replaced.

> Why you report to browser that the file to be delivered is a XLS file, while the real content is CSV?

I was facing problems while writing directly to the XLS file. So, I write into a CSV file and while downloading I changed the content to XLS. The downloaded file is in XLS format. The file is download to the default downloads folder.
export

> Wrong header data

Will fix this now.

Please check this feature on my online server. 
Link: http://stanfordswi.info/engelsystem/public/

**Comment by hongquan** [comment]
I don't ask why you create the file. I concern that what if many users use the export feature and write to that file the same time. "Many users" here is not "non-admin" users, it is "many people login as "admin" role", for example, when H.Phuc and Mario both login to "admin" from different machines and export almost the same time.

An example of problem is that H.Phuc's process want to write "tom@example.com jerry@example.com" to the file, but then Mario's process comes and write "099 654 321", the content of the file will be "tom@example.com 099 654 321 jerry@example.com"

> I was facing problems while writing directly to the XLS file

You don't need to create XLS file.

> while downloading I changed the content to XLS

It is still CSV content:

selection_108

**Comment by hongquan** [comment]
Btw, with your update, the web still write HTML code to HTTP response. But this line

``` php
header("Content-Length: ".filesize($filename));
```

is a trick. It will let browser to cut off the content after the part of CSV content, so that the next part with HTML will be ignored by browser.

You don't need to create XLS file. Every Excel-like softwares can open CSV file. CSV file is also the common format for exporting table data. XLS is a proprietary format and not many software can open it. 

> What do you suggest we should do? 

Use PHP's builtin feature to create temporary file. PHP will generate the file name on it own and we can make sure each user will have her own file, not share with others. Because the file is temporary and created by PHP, PHP will take care of deleting it after use.

**Comment by DishantK1807** [comment]
@hongquan : 

> An example of problem is that H.Phuc's process want to write "tom@example.com jerry@example.com" to the file, but then Mario's process comes and write "099 654 321", the content of the file will be "tom@example.com 099 654 321 jerry@example.com"

got it now. I think in that case, the latest transaction would be saved, there would be some delay in the request between both the servers.
What do you suggest we should do?

**Comment by DishantK1807** [comment]
> You don't need to create XLS file.

Will leave the content as CSV only.

**Comment by hongquan** [comment]
Create temporary file (I updated my comment above)

**Comment by DishantK1807** [comment]
@hongquan: There are problems with making temporary files. If I use temporary files, the exported file that is downloaded is empty because the temporary file is not saved file is not saved. 

can we merge this PR as of now. I'll make a commit resolving the other issues, and create a separate issue for it, and work on it later, as we have other issues to complete too.

**Comment by DishantK1807** [comment]
@hongquan : Updated and tested on online server. Used temporary file as you mentioned.
Kindly review the code.

**Comment by hongquan** [comment]
Please fix the indentation https://github.com/fossasia/engelsystem/pull/68#issuecomment-227042567

You are slowing down the progress of this PR because you continue making mistake in coding style, which has been reminded more than once.

File: includes/pages/admin_import.php [hunk 2]
```diff
@@ -325,49 +337,83 @@ function prepare_events($file, $shifttype_id, $add_minutes_start, $add_minutes_e
...
     );
   return $shifts_printable;
 }
 
 function shift_sort($a, $b) {
   return ($a['start'] < $b['start']) ? - 1 : 1;
 }
+
+function export_xls() {
+	$filename = $fn = tempnam ('/tmp', '.csv'); //  Temporary File Name
```
Comment (hongquan) [inline includes/pages/admin_import.php]: Whitespace `tempnam ('/tmp', '.csv')`.
Is the variable `$fn` used anywhere? If not, delete it.

File: includes/pages/admin_import.php [hunk 3]
```diff
@@ -382,7 +382,7 @@ function shift_sort($a, $b) {
 }
 
 function export_xls() {
-	$filename = $fn = tempnam ('/tmp', '.csv'); //  Temporary File Name
+	$filename = tempnam('/tmp', '.csv');//  Temporary File Name
```
Comment (hongquan) [inline includes/pages/admin_import.php:385]: @DishantK1807 Don't delete the space between the code and the comment. That space is to separate code and comment and make it easier to read the code. In some coding convention, people even require at least 2 spaces here. But Drupal doesn't mention, I will skip it now. But next time, please keep space between code and comment (if those are on the same line).

**Pull request merged**